src/coreclr/nativeaot/Runtime/thread.cpp

```diff
@@ -1341,11 +1341,23 @@ COOP_PINVOKE_HELPER(void, RhpReversePInvokeReturn, (ReversePInvokeFrame * pFrame
 
 #ifdef USE_PORTABLE_HELPERS
 
+#ifdef HOST_WASM
+extern "C" void RhpSetShadowStackTop(void*);
+COOP_PINVOKE_HELPER(void, RhpPInvoke, (void* pShadowStack, PInvokeTransitionFrame* pFrame))
+{
+    // The implementation has to live in this file because of "FORCEINLINE" on "Thread::InlinePInvoke".
+    // Ideally, "InlinePInvoke" would be in "thread.inl", and this function - in "wasm/PInvoke.cpp".
```

Comment:
@VSadov Could you please move these FORCEINLINEs from thread.cpp to thread.inl in one of your cleanup changes?

Reply:
Right. I think we can do that.